endstate_correction/smc.py

```diff
@@ -0,0 +1,146 @@
+        # might help with RBFE calculations in analogy to the work of 10.1101/2020.07.29.227959 and avoid sampling of the
+        # target distribution \pi(\lamb=1; x) for bidirectional NEQ protocols
+
+        # obtain IID samples from the reference distribution \pi(\lamb=0; x)
+        # initialize weights to be uniform
+        # select random subset of walkers from the samples
+        # for each interpolation step i between \lamb=0 and \lamb=1
+        #   propagate the walkers for a fixed number of steps with u(x; \lamb=i) to generate pool of IID samples
+        #   for each walker j
+        #       calculate the unnormalized weights w = exp(u(x; \lamb=i+1) - u(x; \lamb=i))
+        #   resample the walkers according to the normalized weights
+
+        # calculate the free energy estimate using the weights of each walker and the Zwanzig relation
+
+        # initialize weights
+        weights = np.ones(nr_of_walkers) / nr_of_walkers
+        # initialize lambda values
+        lambdas = np.linspace(0, 1, nr_of_steps)
+
+        # select initial samples
+        random_frame_idxs = np.random.choice(
+            len(self.samples.xyz) - 1, size=nr_of_walkers
+        )
+        walkers = [
+            self.samples.openmm_positions(random_frame_idx)
+            for random_frame_idx in random_frame_idxs
+        ]
+
+        assert len(walkers) == nr_of_walkers
+
+        # start with switch
+        for lamb_idx in tqdm(range(len(lambdas) - 1)):
+            # set lambda parameter
+            self.sim.context.setParameter("lambda_interpolate", lambdas[lamb_idx])
+            # Propagate the walkers
+            walkers = self._propagate_particles(walkers, self.sim)
+            # calculate work
+            # evaluate U_(\lamb_(i+1))(x_i) -  U_(\lamb_(i))(x_i)
+            # calculate U_(\lamb_(i))(x_i)
+            u_now = self._calculate_potential_E_for_particles(
+                lambdas[lamb_idx], walkers, self.sim
+            )
+            # calculate U_(\lamb_(i+1))(x_i)
+            u_future = self._calculate_potential_E_for_particles(
+                lambdas[lamb_idx + 1], walkers, self.sim
+            )
+            # calculate weights (equation 2 in 10.1021/acs.jctc.1c01198)
+            current_deltaEs = u_future - u_now  # calculate difference in energy
+            current_weights = np.exp(
+                np.nanmin(current_deltaEs) - current_deltaEs
+            )  # subtract minimum to avoid numerical issues
+            current_weights /= np.sum(current_weights)  # normalize
+            # calculate ESS
+            ESS = 1.0 / np.sum(current_weights**2)
+            print(f"Effective Sample Size at lambda = {lambdas[lamb_idx]}: {ESS}")
+            print(current_weights)
+            # add to accumulated logZ
+            self.logZ += logsumexp(-current_deltaEs) - np.log(nr_of_walkers)
+
+            # Resample the particles based on the weights
+            random_frame_idxs = np.random.choice(
```

Comment:
Ideally it's best to use stratified resampling because this is maximally representative of your weights. With regular random choice you are adding sampling noise for no reason. So would be good to abstract this as well.

E.g. if you have 2 samples each with weight 0.5, stratified resampling would always resample them such that you preserve both of these samples. Random choice has a 50% chance that it will pick the same sample twice which is not great.

Reply:
great point! I took inspiration from your implementation and added a class and method for it